### PR TITLE
Use functions to query stages and stage masks

### PR DESCRIPTION
**Pull request by kuhar**
This introduces new helper function to `llpcUtils.h` to replace bug-prone
and verbose manual bit manipulation.

The new functions should make it more immediately obvious what the intention
behind each check is. For example, there are some subtleties around what
to do with copy shader stages, which stages should be graphics vs. native, etc.
This PR attempts to make it clearer and more consistent; each new function
should have well-defined semantics, documented via unit tests.
    
Also, introduce a helper function to get all shader stages in a mask to make
loops more concise.

### CODE REVIEW

**Comment by kuhar** [comment]
This PR depends on the changes in https://github.com/GPUOpen-Drivers/llpc/pull/1574.
Only the second commit 'Use function to query...' should be reviewed.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5848808e1a42d61e8b6ddadab4b4f4727f88905c
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 44abc48d199dd28369bf7415eee1751c4bf616d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev b8ed024d536d1bebb35927a0f4f470dca2f2dfc6</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev cdd74831d061bf65fdf5af4a0b6ce99fe2c558a9</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/163959)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32354/54653 (59.2%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22299/54653 (40.8%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32530/54653 (59.5%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22123/54653 (40.5%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35511/54653 (65.0%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 19142/54653 (35.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
V1: Rebase (#1574 got merged), no code changes.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 395f79c728dbe2c376df49c623de89b5b2ccb1ad
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 44abc48d199dd28369bf7415eee1751c4bf616d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev b8ed024d536d1bebb35927a0f4f470dca2f2dfc6</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev cdd74831d061bf65fdf5af4a0b6ce99fe2c558a9</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/163959)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32354/54653 (59.2%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22299/54653 (40.8%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32530/54653 (59.5%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22123/54653 (40.5%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35511/54653 (65.0%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 19142/54653 (35.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
Ping for a code review. @amdrexu @dstutt @vettoreldaniele

File: llpc/unittests/util/testUtil.cpp
Comment (vettoreldaniele) [inline llpc/unittests/util/testUtil.cpp:169]: nit: typo `is` -> `it`

File: llpc/unittests/util/testUtil.cpp
Comment (vettoreldaniele) [inline llpc/unittests/util/testUtil.cpp:75]: nit: typo is -> it

File: llpc/unittests/util/testUtil.cpp
Comment (vettoreldaniele) [inline llpc/unittests/util/testUtil.cpp:43]: Do we need this placeholder? What's its purpose?

File: llpc/tool/llpcCompilationUtils.cpp
Comment (vettoreldaniele) [inline llpc/tool/llpcCompilationUtils.cpp:494]: Slightly unrelated to the change but it caught my attention, should we assert that there's only a single stage here?

File: llpc/context/llpcPipelineContext.cpp
Comment (vettoreldaniele) [inline llpc/context/llpcPipelineContext.cpp:206]: Can you use `make_filter_range(allStages, isNativeStage)` here as well?

**Review by vettoreldaniele — CHANGES_REQUESTED** [review]

File: llpc/unittests/util/testUtil.cpp
Comment (kuhar) [reply llpc/unittests/util/testUtil.cpp:43]: I use as the simplest possible test that's trivially correct. Then when I suppress cppcheck errors it's clear that the cppcheck is at fault and not the test. But in this case, it does not appear to work because I had to annotate the next one too.

File: llpc/tool/llpcCompilationUtils.cpp
Comment (kuhar) [reply llpc/tool/llpcCompilationUtils.cpp:494]: See the if above: `if (stages.empty()) return ...`

**Comment by kuhar** [comment]
V2: Apply @vettoreldaniele 's suggestions. Rebase.

File: llpc/unittests/util/testUtil.cpp
Comment (kuhar) [reply llpc/unittests/util/testUtil.cpp:43]: I had to re-add it to make cppckeck happy

**Comment by kuhar** [comment]
V3. Fix test name.

**Comment by amdvlk-admin** [comment]
### Test summary for commit b95576c9046510dafdecd4ce699c1f297504e94f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 44abc48d199dd28369bf7415eee1751c4bf616d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev b8ed024d536d1bebb35927a0f4f470dca2f2dfc6</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev cdd74831d061bf65fdf5af4a0b6ce99fe2c558a9</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/218612)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32354/54653 (59.2%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22299/54653 (40.8%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 32194/54653 (58.9%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22458/54653 (41.1%)</li><li>Warnings: 1/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32530/54653 (59.5%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22123/54653 (40.5%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35511/54653 (65.0%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 19142/54653 (35.0%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit d74e48353bfe05a2c10f99275d016ad02e4f5f7b
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 44abc48d199dd28369bf7415eee1751c4bf616d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev b8ed024d536d1bebb35927a0f4f470dca2f2dfc6</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev cdd74831d061bf65fdf5af4a0b6ce99fe2c558a9</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/209382)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32354/54653 (59.2%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22299/54653 (40.8%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 29693/45423 (65.4%)</li>
<li><details><summary>Failed: 1/45423 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.clustered.graphics.subgroupclusteredmin_u8vec3
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 15728/45423 (34.6%)</li><li>Warnings: 1/45423 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32530/54653 (59.5%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22123/54653 (40.5%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35511/54653 (65.0%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 19142/54653 (35.0%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/util/llpcUtil.h
Comment (amdrexu) [inline llpc/util/llpcUtil.h:134]: Please add separate line "// ====" for each function according to the coding style even if they are defined in header file.

**Review by amdrexu — COMMENTED** [review]
LGTM.
It's a good change. But we are still processing the PR https://github.com/GPUOpen-Drivers/llpc/pull/1535 in our internal repo to add additional codes. Maybe this change will wait for some time to be merged. I think after it is merged we still need to add extra codes in our internal repo. Just heads-up and Cc @JaxLinAMD

File: llpc/util/llpcUtil.h
Comment (kuhar) [reply llpc/util/llpcUtil.h:135]: Done

**Comment by kuhar** [comment]
V4. Add missing documentation to `llpcUtil.h`. Rebase.

**Comment by kuhar** [comment]
> LGTM. It's a good change. But we are still processing the PR #1535 in our internal repo to add additional codes. Maybe this change will wait for some time to be merged. I think after it is merged we still need to add extra codes in our internal repo. Just heads-up and Cc @JaxLinAMD

@amdrexu OK, thanks. If you can, please prioritize merging this PR, as I have more changes stacked on top of it.

File: llpc/tool/llpcCompilationUtils.cpp
Comment (kuhar) [reply llpc/tool/llpcCompilationUtils.cpp:493]: Yes, that's exactly what happens in the code. We always pick the first one, as ordered by the list of stages in the enum definition (`Vkgc::ShaderStage`). This change preserves the existing logic. Before, the loop would break after finding the first stage.

In this PR, I want to preserve the existing logic and not do any unrelated clean ups if I can avoid them.

Comment (kuhar) [reply llpc/tool/llpcCompilationUtils.cpp:493]: And an assertion like you suggest fails in LIT tests. There is sometimes more than one stage.

**Comment by kuhar** [comment]
retest this please

**Review by vettoreldaniele — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 01555aa62a521aea58756646300b2623201132d5
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 44abc48d199dd28369bf7415eee1751c4bf616d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev b8ed024d536d1bebb35927a0f4f470dca2f2dfc6</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev cdd74831d061bf65fdf5af4a0b6ce99fe2c558a9</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/218612)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 32354/54653 (59.2%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22299/54653 (40.8%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 32194/54653 (58.9%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22458/54653 (41.1%)</li><li>Warnings: 1/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 32530/54653 (59.5%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 22123/54653 (40.5%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35511/54653 (65.0%)</li>
<li>Failed: 0/54653 (0.0%)</li>
<li>Not Supported: 19142/54653 (35.0%)</li><li>Warnings: 0/54653 (0.0%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Pull request merged**
